### PR TITLE
[4] Revise Polygon clip testing condition

### LINKED ISSUE #4: Clipped polygon is incorrect
The progam clips/cuts the image for volume calculating based on a WKT (well-known-text) polygon.
However, testing results point out that the clipped region is imprecise.

The test result (clipped images) is in the link below:
https://docs.google.com/spreadsheets/d/1J3nusoqYMRisPixA7XqrSIiFvFcaA1VNUqeV3GigcmU/edit?usp=sharing

### PR DESCRIPTION
**Pull request by lhviet**
## Description
This PR revise the clip testing condition by a function added in `extent_utils.py` to solve cutting issue described in issue #4.
Along the way, few parameters added to allow (or not) the clipped polygons saved as PNG/TIFF to the same directory of the DSM files for testing. The clips will be saved in default.

Polygon:
fireshot capture 55 - angelswing - https___dev ondronemap com_project_118_content

Current:
clipped-image-lake

Result after this PR:
clipped-image-1140-lake

Resolves #4

### CODE REVIEW

File: src/extent_utils.py [hunk 1]
```diff
@@ -40,3 +40,35 @@ def apply_gt(geo_trans, ext, from_dic=True, to_dic=True):
     tup = (min_x, max_x, min_y, max_y)
 
     return tup_to_dic(tup) if to_dic else tup
+
+def revise_geom_pxl_point(x, y, ext_bound_x, ext_bound_y):
+    """
+        During converting, the x, y value may bigger than ext max_x, max_y values
+        few units (in practice, just 1px, expecting < 5px).
```
Comment (Ailrun) [inline src/extent_utils.py]: It's not just bigger than few units, but also  can be much bigger `ext_bound_x`, since user can select the point outside of the tiff bound.

File: src/extent_utils.py [hunk 2]
```diff
@@ -40,3 +40,35 @@ def apply_gt(geo_trans, ext, from_dic=True, to_dic=True):
...
+            return  # FIXME: We need to handle this case also
+    if 0 <= y < ext_bound_y:    # if point y is inside of the extent, it is fine
+        revised_y = y
+    else:
+        if y <= ext_bound_y + extended_unit:    # if point y is outside of the extent, using the bound point y
+            revised_y = y - 1
+        else:
+            return  # FIXME: We need to handle this case also
+    if revised_x is not None and revised_y is not None:
+        return tuple((revised_x, revised_y))
```
Comment (Ailrun) [inline src/extent_utils.py]: Line feed is missed.

File: src/calculate_volume.py [hunk 3]
```diff
@@ -7,7 +7,7 @@
 
 import clip as c
 
-def calculate_volume(tiff, geom):
+def calculate_volume(tiff, geom, clip_save_name='', isSaveClipInPNG=True):
```
Comment (Ailrun) [inline src/calculate_volume.py]: 1. These `clip_save_name` and `isSaveClipInPNG` use mixed naming convention.
1. These are only for testing, so default should be `False`.

File: src/clip.py [hunk 4]
```diff
@@ -85,21 +87,21 @@ def clip(tiff_gt, tiff_arr, geom, no_data=-10000):
...
         return np.array([]), []
 
-    raster_size = (
-        int(geom_pxl_ext['max_x'] - geom_pxl_ext['min_x']),
-        int(geom_pxl_ext['max_y'] - geom_pxl_ext['min_y']),
-    )
+    raster_size = (ext_bound_x, ext_bound_y)
+
     raster_poly = Image.new('L', raster_size, color=0x000001)
+
```
Comment (Ailrun) [inline src/clip.py]: Why did you add this linefeed? I mean, line 103 and line 105 is about setting `raster_poly` with appropriate initial value.

File: src/extent_utils.py (hunk 1 above)
Comment (lhviet) [reply src/extent_utils.py]: This is not to solve that case, out of tiff bound, but solve the case mis-calculation and the points out of **extent bound**

File: src/extent_utils.py (hunk 2 above)
Comment (lhviet) [reply src/extent_utils.py]: Do you mean line 72, need a blank line after return?

File: src/extent_utils.py (hunk 1 above)
Comment (Ailrun) [reply src/extent_utils.py]: I cannot understand what you mean.
Extend bound should include all points of WKT if all points are in tiff extend. It can go out extend bound if and only if point goes out of tiff extend

File: src/extent_utils.py (hunk 2 above)
Comment (Ailrun) [reply src/extent_utils.py]: I mean the final one. According to POSIX standard, you should put linefeed at the end of the file.

File: src/extent_utils.py (hunk 1 above)
Comment (lhviet) [reply src/extent_utils.py]: Yes, that is what we are expecting, `Extend bound should include all points of WKT if all points are in tiff extend`. However, from my test, a high percentage of possibility that 1 point will not be included in the extent.
The reason is, the previous comparison that I mentioned in the comments (in code):
```
clipped_pxls = [
        (x, y)
        for (x, y) in pxls
        if 0 <= x < geom_pxl_ext['max_x'] - geom_pxl_ext['min_x'] and
        0 <= y < geom_pxl_ext['max_y'] - geom_pxl_ext['min_y']
    ]
```
In my test, i.e., 
`geom_pxl_ext['max_x'] - geom_pxl_ext['min_x']` is **60**
and we expect `x` to always in range [0, 59].
However, the practice is `x` may be equal to **60**, and it happens frequently. The same to `y`.

File: src/calculate_volume.py (hunk 3 above)
Comment (lhviet) [reply src/calculate_volume.py]: 2. Actually to decide to save the clipped polygon to image or not is in `clip_save_name=''`, if no name set (default), there is no addition process like saving the clips.
`isSaveClipInPNG=True` is used to control the saving extension, **PNG** or **TIFF**.
1. The variable will be renamed `isSaveClipInPNG --> is_clip_saved_in_png`

File: src/extent_utils.py (hunk 1 above)
Comment (Ailrun) [reply src/extent_utils.py]: @lhviet Then you can fix it to use `<=`, couldn't you?

Comment (Ailrun) [reply src/extent_utils.py]: I mean, test says that both `max` and `min` are inclusive things, so you can fix it with `<=`s.

Comment (lhviet) [reply src/extent_utils.py]: @Ailrun you got the point.
`<=` will make it simple.
The reason why I fix by `+5` instead of `<=` is to prevent the case it is out of the extent within few units.
Secondly, if using `<=`, the point still outside of the extent, and will make side effect in later calculation, instead, we need to decrease the point (x, or y) to be inside of the extent.
```
    if 0 <= x < ext_bound_x:    # if point x is inside of the extent, it is fine
        revised_x = x
    else:
        if x == ext_bound_x + extended_unit:    # if point x is outside of the extent, using the bound point x
            revised_x = ext_bound_x  - 1
        else:
            return
```

Comment (Ailrun) [reply src/extent_utils.py]: @lhviet To me,
> using <=, the point still outside of the extent
Cannot make sense... and what side effect? `<=` does not have any side effect. If max and min values are both inclusive, other code is wrong so that **they should be fixed**.

Comment (Ailrun) [reply src/extent_utils.py]: And if they're inclusive, you code `ext_bound_x  - 1` will also lead to wrong result (it will move boundary 1 pixel inner)

Comment (lhviet) [reply src/extent_utils.py]: @Ailrun 
Here for example, we have the extent with 
`width (boundX) = 60px`, where `minX = 0` and `maxX = 59`
We expect point with x is in range of `[0, 59]`, but using `x <= 60`, we will have a point with `x = 60`.
In this case, revise x by using `revised_x = boundX - 1 = 59` will solve the problem.

Comment (lhviet) [reply src/extent_utils.py]: I do not move the boundary, I revised the point coordinate to be inside the extent. `x` is replaced by `revised_x`, and `revised_x = ext_bound_x  - 1`

Comment (Ailrun) [reply src/extent_utils.py]: ? if maxX and minX is 0 and 59, the expression should be `x <= 59`, not `x <= 60`.

Comment (Ailrun) [reply src/extent_utils.py]: Currently, our expression is `x < 59`.

Comment (Ailrun) [reply src/extent_utils.py]: @lhviet It will move the boundary, since maxX is inclusive, `x = maxX` is valid point in the extent, and you revise it to `x = maxX - 1`.

Comment (lhviet) [reply src/extent_utils.py]: Sorry, lets have a better real example (the lake polygon in my description):

This is the Extent: `geom_pxl_ext = {'min_x': 152, 'max_x': 437, 'min_y': 674, 'max_y': 989}`
We will have the bound 
```
ext_bound_x = 285
ext_bound_y = 315
```
And this is the polygon `pxls = [(0, 16), (54, 0), (91, 48), (180, 7), (284, 70), (159, 199), (108, 261), (92, 315), (47, 302), (40, 252), (22, 243), (4, 169), (20, 156)]`

**(92, 315)** will be removed in this case, and certainly we will have a wrong result here.

File: src/clip.py [hunk 5]
```diff
@@ -85,20 +87,18 @@ def clip(tiff_gt, tiff_arr, geom, no_data=-10000):
...
+            clipped_pxls.append(polygon_pt)
 
     if len(clipped_pxls) < 2:
         return np.array([]), []
 
-    raster_size = (
-        int(geom_pxl_ext['max_x'] - geom_pxl_ext['min_x']),
-        int(geom_pxl_ext['max_y'] - geom_pxl_ext['min_y']),
-    )
+    raster_size = (ext_bound_x, ext_bound_y)
```
Comment (Ailrun) [inline src/clip.py]: This is the point where calculation goes wrong,
since real `raster_size` is `ext_bound_x + 1, ext_bound_y + 1` because max and min are inclusive.

Comment (Ailrun) [reply src/clip.py]: Line 69-72 also can goes wrong, since it assume that max is exclusive.

File: src/extent_utils.py (hunk 1 above)
Comment (Ailrun) [reply src/extent_utils.py]: @lhviet Which will not be removed with `y <= ext_bound_y`, and that's what I said.

Comment (Ailrun) [reply src/extent_utils.py]: However, it cannot be drawn to file because of https://github.com/angelswing-team/volume-calculator-for-dsm/pull/5#discussion_r210788098, but it's the problem of other codes.

File: src/clip.py (hunk 5 above)
Comment (lhviet) [reply src/clip.py]: @Ailrun Line 58-61 & 69-72 also can go wrong, but will not included in this fix because I have no good solution now.

Why `raster_size = (ext_bound_x, ext_bound_y)` can go wrong?
Line 90-91 like below, and there is no modification assign to `ext_bound_x, ext_bound_y`
```
    ext_bound_x = geom_pxl_ext['max_x'] - geom_pxl_ext['min_x']
    ext_bound_y = geom_pxl_ext['max_y'] - geom_pxl_ext['min_y']
```

Comment (Ailrun) [reply src/clip.py]: @lhviet Since max and min are both inclusive.
Let me give an example.
When `max_x = 3` and `min_x = 1` and both are inclusive, we can have point on `x = 1` and `x = 3`. So, what we can have are `x = 1`, `x = 2`, `x = 3`, those three cases.
But when we calculate width with `max_x - min_x`, then the result is `2`, and usually, pixel width 2 means there is only two pixels (where we need three pixels) in the one row.

File: src/extent_utils.py (hunk 1 above)
Comment (lhviet) [reply src/extent_utils.py]: @Ailrun 
Sure `y <= ext_bound_y` will not remove the point.
But as I said, the point is out of the bound of the extent.
And the following code, line 101-103, I cannot sure if any effect. Do you still want to use `<=`
```
    raster_size = (ext_bound_x, ext_bound_y)
    raster_poly = Image.new('L', raster_size, color=0x000001)
    ImageDraw.Draw(raster_poly).polygon(clipped_pxls, fill=0x000000)
```

Comment (Ailrun) [reply src/extent_utils.py]: @lhviet Fixing `<=`, raster size, and Line 58-61 and Line 69-72 should solve problem related with inclusive `max`. There may exists other codes using assumption of exclusive `max`, though...

File: src/clip.py (hunk 5 above)
Comment (lhviet) [reply src/clip.py]: @Ailrun 
Will this fix (line 58) ?
```
        if geom_pxl_ext['max_y'] - 1 > tiff_shape[1]:
            geom_pxl_ext['max_y'] = tiff_shape[1]
```

Comment (Ailrun) [reply src/clip.py]: @lhviet It will still give you an index outside of tiff. It should be
```python
if geom_pxl_ext['max_y'] > tiff_shape[1] - 1:
  geom_pxl_ext['max_y'] = tiff_shape[1] - 1
```

Comment (lhviet) [reply src/clip.py]: Ah, I got it. Thanks.

Comment (Ailrun) [reply src/clip.py]: I really hate all thoes inclusive/exclusive thingies... :angry:

File: src/clip.py [hunk 6]
```diff
@@ -53,21 +55,25 @@ def clip(tiff_gt, tiff_arr, geom, no_data=-10000):
         geom_pxl_ext['max_x'] = 0
 
     if len(tiff_shape) == 3:
-        if geom_pxl_ext['max_y'] > tiff_shape[1]:
-            geom_pxl_ext['max_y'] = tiff_shape[1]
-        if geom_pxl_ext['max_x'] > tiff_shape[2]:
-            geom_pxl_ext['max_x'] = tiff_shape[2]
+        bound_tiff_y = tiff_shape[1] - 1
+        bound_tiff_x = tiff_shape[2] - 1
```
Comment (Ailrun) [inline src/clip.py]: How about define those before `if`? they are used in both `if` and `else`.

File: src/clip.py (hunk 5 above)
Comment (lhviet) [reply src/clip.py]: Yes, and it take time for me to print out polygons to images to figure out where the problem from.
Anw, I updated, please review again.

File: src/clip.py (hunk 6 above)
Comment (lhviet) [reply src/clip.py]: But, inside if, they have different index.

Comment (Ailrun) [reply src/clip.py]: @lhviet Oh, you're right. I miss that.
